diff --git a/client/Windows/wfreerdp.rc b/client/Windows/wfreerdp.rc
index cec908f46..8446477b3 100644
--- a/client/Windows/wfreerdp.rc
+++ b/client/Windows/wfreerdp.rc
@@ -2,13 +2,13 @@
 #include "resource.h"
 
 IDI_ICON1              ICON                    "FreeRDP.ico"
-IDB_MINIMIZE           BITMAP                  "resource\\minimize.bmp"
-IDB_MINIMIZE_ACT       BITMAP                  "resource\\minimize_active.bmp"
-IDB_LOCK               BITMAP                  "resource\\lock.bmp"
-IDB_LOCK_ACT           BITMAP                  "resource\\lock_active.bmp"
-IDB_UNLOCK             BITMAP                  "resource\\unlock.bmp"
-IDB_UNLOCK_ACT         BITMAP                  "resource\\unlock_active.bmp"
-IDB_CLOSE              BITMAP                  "resource\\close.bmp"
-IDB_CLOSE_ACT          BITMAP                  "resource\\close_active.bmp"
-IDB_RESTORE            BITMAP                  "resource\\restore.bmp"
-IDB_RESTORE_ACT        BITMAP                  "resource\\restore_active.bmp"
+IDB_MINIMIZE           BITMAP                  "resource/minimize.bmp"
+IDB_MINIMIZE_ACT       BITMAP                  "resource/minimize_active.bmp"
+IDB_LOCK               BITMAP                  "resource/lock.bmp"
+IDB_LOCK_ACT           BITMAP                  "resource/lock_active.bmp"
+IDB_UNLOCK             BITMAP                  "resource/unlock.bmp"
+IDB_UNLOCK_ACT         BITMAP                  "resource/unlock_active.bmp"
+IDB_CLOSE              BITMAP                  "resource/close.bmp"
+IDB_CLOSE_ACT          BITMAP                  "resource/close_active.bmp"
+IDB_RESTORE            BITMAP                  "resource/restore.bmp"
+IDB_RESTORE_ACT        BITMAP                  "resource/restore_active.bmp"
diff --git a/winpr/include/winpr/spec.h b/winpr/include/winpr/spec.h
index d1a2d740a..1c5a86451 100644
--- a/winpr/include/winpr/spec.h
+++ b/winpr/include/winpr/spec.h
@@ -965,6 +965,7 @@ extern "C++"
 
 #endif
 
+#ifndef __MINGW32__
 #if defined(_WIN32) || defined(__CYGWIN__)
 #ifdef __GNUC__
 #define DECLSPEC_EXPORT __attribute__((dllexport))
@@ -982,5 +983,6 @@ extern "C++"
 #define DECLSPEC_IMPORT
 #endif
 #endif
+#endif
 
 #endif /* WINPR_SPEC_H */
diff --git a/winpr/include/winpr/wtypes.h.in b/winpr/include/winpr/wtypes.h.in
index e4ee27a5b..d05c3ca48 100644
--- a/winpr/include/winpr/wtypes.h.in
+++ b/winpr/include/winpr/wtypes.h.in
@@ -578,4 +578,9 @@ typedef const BYTE* LPCBYTE;
 #pragma clang diagnostic pop
 #endif
 
+#ifdef __MINGW32__
+#undef interface
+#define WAVE_MAPPED_DEFAULT_COMMUNICATION_DEVICE (0x0010)
+#endif
+
 #endif /* WINPR_WTYPES_H */
diff --git a/winpr/libwinpr/sspi/NTLM/ntlm.h b/winpr/libwinpr/sspi/NTLM/ntlm.h
index 21ae893a3..8ec8fce87 100644
--- a/winpr/libwinpr/sspi/NTLM/ntlm.h
+++ b/winpr/libwinpr/sspi/NTLM/ntlm.h
@@ -77,6 +77,7 @@ enum _NTLM_STATE
 };
 typedef enum _NTLM_STATE NTLM_STATE;
 
+#ifndef __MINGW32__
 enum _NTLM_AV_ID
 {
 	MsvAvEOL,
@@ -91,6 +92,16 @@ enum _NTLM_AV_ID
 	MsvAvTargetName,
 	MsvAvChannelBindings
 };
+#else
+enum _NTLM_AV_ID
+{
+        MinGw_NTLM_AV_ID_0
+};
+#define MsvAvTimestamp (MsvAvFlags + 1)
+#define MsvAvSingleHost (MsvAvTimestamp + 1)
+#define MsvAvTargetName (MsvAvSingleHost + 1)
+#define MsvAvChannelBindings (MsvAvTargetName + 1)
+#endif
 typedef enum _NTLM_AV_ID NTLM_AV_ID;
 
 struct _NTLM_AV_PAIR
diff --git a/winpr/libwinpr/wtsapi/wtsapi.c b/winpr/libwinpr/wtsapi/wtsapi.c
index 04c59f9fa..24da90ada 100644
--- a/winpr/libwinpr/wtsapi/wtsapi.c
+++ b/winpr/libwinpr/wtsapi/wtsapi.c
@@ -53,7 +53,7 @@ static HMODULE g_WtsApi32Module = NULL;
 static WtsApiFunctionTable WtsApi32_WtsApiFunctionTable = { 0 };
 
 #define WTSAPI32_LOAD_PROC(_name, _type)                                                   \
-	WtsApi32_WtsApiFunctionTable.p##_name = (##_type)GetProcAddress(g_WtsApi32Module, "WT" \
+	WtsApi32_WtsApiFunctionTable.p##_name = (_type)GetProcAddress(g_WtsApi32Module, "WT" \
 	                                                                                  "S" #_name);
 
 static BOOL WtsApi32_InitializeWtsApi(void)
diff --git a/winpr/libwinpr/CMakeLists.txt b/winpr/libwinpr/CMakeLists.txt
index 787120248..96e31a7b2 100644
--- a/winpr/libwinpr/CMakeLists.txt
+++ b/winpr/libwinpr/CMakeLists.txt
@@ -144,6 +144,9 @@ if (WITH_LIBRARY_VERSIONING)
 endif()
 
 add_definitions(${WINPR_DEFINITIONS})
+if (MINGW)
+  winpr_library_add_public(msvcr120 ws2_32 mfplat strmiids cfgmgr32)
+endif()
 target_link_libraries(${MODULE_NAME} PRIVATE ${WINPR_LIBS_PRIVATE} PUBLIC ${WINPR_LIBS_PUBLIC})
 install(TARGETS ${MODULE_NAME} COMPONENT libraries EXPORT WinPRTargets
         ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
