diff --git a/client/Windows/wfreerdp.rc b/client/Windows/wfreerdp.rc
index cec908f46..8446477b3 100644
--- a/client/Windows/wfreerdp.rc
+++ b/client/Windows/wfreerdp.rc
@@ -2,13 +2,13 @@
 #include "resource.h"
 
 IDI_ICON1              ICON                    "FreeRDP.ico"
-IDB_MINIMIZE           BITMAP                  "resource\\minimize.bmp"
-IDB_MINIMIZE_ACT       BITMAP                  "resource\\minimize_active.bmp"
-IDB_LOCK               BITMAP                  "resource\\lock.bmp"
-IDB_LOCK_ACT           BITMAP                  "resource\\lock_active.bmp"
-IDB_UNLOCK             BITMAP                  "resource\\unlock.bmp"
-IDB_UNLOCK_ACT         BITMAP                  "resource\\unlock_active.bmp"
-IDB_CLOSE              BITMAP                  "resource\\close.bmp"
-IDB_CLOSE_ACT          BITMAP                  "resource\\close_active.bmp"
-IDB_RESTORE            BITMAP                  "resource\\restore.bmp"
-IDB_RESTORE_ACT        BITMAP                  "resource\\restore_active.bmp"
+IDB_MINIMIZE           BITMAP                  "resource/minimize.bmp"
+IDB_MINIMIZE_ACT       BITMAP                  "resource/minimize_active.bmp"
+IDB_LOCK               BITMAP                  "resource/lock.bmp"
+IDB_LOCK_ACT           BITMAP                  "resource/lock_active.bmp"
+IDB_UNLOCK             BITMAP                  "resource/unlock.bmp"
+IDB_UNLOCK_ACT         BITMAP                  "resource/unlock_active.bmp"
+IDB_CLOSE              BITMAP                  "resource/close.bmp"
+IDB_CLOSE_ACT          BITMAP                  "resource/close_active.bmp"
+IDB_RESTORE            BITMAP                  "resource/restore.bmp"
+IDB_RESTORE_ACT        BITMAP                  "resource/restore_active.bmp"
diff --git a/winpr/include/winpr/crt.h b/winpr/include/winpr/crt.h
index 66aa4a79b..0b531a5c0 100644
--- a/winpr/include/winpr/crt.h
+++ b/winpr/include/winpr/crt.h
@@ -165,6 +165,13 @@ extern "C"
 	WINPR_API errno_t wmemmove_s(WCHAR* dest, size_t numberOfElements, const WCHAR* src,
 	                             size_t count);
 
+#endif
+#ifdef __MINGW32__
+	WINPR_API void* _mingw_aligned_recalloc(void* memblock, size_t num, size_t size, size_t alignment);
+#define _aligned_recalloc(MEMBLOCK,NUM,SIZE,ALIGNMENT) _mingw_aligned_recalloc(MEMBLOCK,NUM,SIZE,ALIGNMENT)
+#endif
+#ifdef _WIN32
+
 #ifdef __cplusplus
 }
 #endif
diff --git a/winpr/include/winpr/intrin.h b/winpr/include/winpr/intrin.h
index b6c3a48ce..1be85f7d5 100644
--- a/winpr/include/winpr/intrin.h
+++ b/winpr/include/winpr/intrin.h
@@ -31,7 +31,10 @@
  * Beware: the result of __builtin_clz(0) is undefined
  */
 
-#if (__GNUC__ > 4) || ((__GNUC__ == 4) && (__GNUC_MINOR__ >= 2))
+#if defined(__MINGW32__) && !defined(__ARM_ARCH)
+#define __DEFAULT_FN_ATTRS __attribute__((__nodebug__, __target__("lzcnt")))
+static unsigned int __DEFAULT_FN_ATTRS __lzcnt(unsigned int __X) { return __builtin_ia32_lzcnt_u32(__X); }
+#elif (__GNUC__ > 4) || ((__GNUC__ == 4) && (__GNUC_MINOR__ >= 2))
 
 static INLINE UINT32 __lzcnt(UINT32 _val32)
 {
diff --git a/winpr/include/winpr/spec.h b/winpr/include/winpr/spec.h
index d1a2d740a..1c5a86451 100644
--- a/winpr/include/winpr/spec.h
+++ b/winpr/include/winpr/spec.h
@@ -965,6 +965,7 @@ extern "C++"
 
 #endif
 
+#ifndef __MINGW32__
 #if defined(_WIN32) || defined(__CYGWIN__)
 #ifdef __GNUC__
 #define DECLSPEC_EXPORT __attribute__((dllexport))
@@ -982,5 +983,6 @@ extern "C++"
 #define DECLSPEC_IMPORT
 #endif
 #endif
+#endif
 
 #endif /* WINPR_SPEC_H */
diff --git a/winpr/include/winpr/wtypes.h.in b/winpr/include/winpr/wtypes.h.in
index e4ee27a5b..8f35a9e11 100644
--- a/winpr/include/winpr/wtypes.h.in
+++ b/winpr/include/winpr/wtypes.h.in
@@ -578,4 +578,9 @@ typedef const BYTE* LPCBYTE;
 #pragma clang diagnostic pop
 #endif
 
+#ifdef __MINGW32__
+#undef interface
+#define WAVE_MAPPED_DEFAULT_COMMUNICATION_DEVICE (0x0010)
+#endif
+
 #endif /* WINPR_WTYPES_H */
diff --git a/winpr/libwinpr/CMakeLists.txt b/winpr/libwinpr/CMakeLists.txt
index 787120248..36a1fdadc 100644
--- a/winpr/libwinpr/CMakeLists.txt
+++ b/winpr/libwinpr/CMakeLists.txt
@@ -144,6 +144,9 @@ if (WITH_LIBRARY_VERSIONING)
 endif()
 
 add_definitions(${WINPR_DEFINITIONS})
+if (MINGW)
+  winpr_library_add_public(stdc++ ws2_32 mfplat strmiids cfgmgr32)
+endif()
 target_link_libraries(${MODULE_NAME} PRIVATE ${WINPR_LIBS_PRIVATE} PUBLIC ${WINPR_LIBS_PUBLIC})
 install(TARGETS ${MODULE_NAME} COMPONENT libraries EXPORT WinPRTargets
         ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
diff --git a/winpr/libwinpr/crt/alignment.c b/winpr/libwinpr/crt/alignment.c
index 6d45d2d58..548898ef7 100644
--- a/winpr/libwinpr/crt/alignment.c
+++ b/winpr/libwinpr/crt/alignment.c
@@ -48,6 +48,16 @@
 #include "../log.h"
 #define TAG WINPR_TAG("crt")
 
+#endif
+#ifdef __MINGW32__
+void* _mingw_aligned_recalloc(void* memblock, size_t num, size_t size, size_t alignment)
+{
+       void* ptr = _aligned_offset_realloc(memblock, size, alignment, 0);
+       return ptr;
+}
+#endif
+#ifndef _WIN32
+
 struct winpr_aligned_mem
 {
 	UINT32 sig;
diff --git a/winpr/libwinpr/sspi/NTLM/ntlm.h b/winpr/libwinpr/sspi/NTLM/ntlm.h
index 21ae893a3..8ec8fce87 100644
--- a/winpr/libwinpr/sspi/NTLM/ntlm.h
+++ b/winpr/libwinpr/sspi/NTLM/ntlm.h
@@ -77,6 +77,7 @@ enum _NTLM_STATE
 };
 typedef enum _NTLM_STATE NTLM_STATE;
 
+#ifndef __MINGW32__
 enum _NTLM_AV_ID
 {
 	MsvAvEOL,
@@ -91,6 +92,16 @@ enum _NTLM_AV_ID
 	MsvAvTargetName,
 	MsvAvChannelBindings
 };
+#else
+enum _NTLM_AV_ID
+{
+        MinGw_NTLM_AV_ID_0
+};
+#define MsvAvTimestamp (MsvAvFlags + 1)
+#define MsvAvSingleHost (MsvAvTimestamp + 1)
+#define MsvAvTargetName (MsvAvSingleHost + 1)
+#define MsvAvChannelBindings (MsvAvTargetName + 1)
+#endif
 typedef enum _NTLM_AV_ID NTLM_AV_ID;
 
 struct _NTLM_AV_PAIR
diff --git a/winpr/libwinpr/wtsapi/wtsapi.c b/winpr/libwinpr/wtsapi/wtsapi.c
index 04c59f9fa..24da90ada 100644
--- a/winpr/libwinpr/wtsapi/wtsapi.c
+++ b/winpr/libwinpr/wtsapi/wtsapi.c
@@ -53,7 +53,7 @@ static HMODULE g_WtsApi32Module = NULL;
 static WtsApiFunctionTable WtsApi32_WtsApiFunctionTable = { 0 };
 
 #define WTSAPI32_LOAD_PROC(_name, _type)                                                   \
-	WtsApi32_WtsApiFunctionTable.p##_name = (##_type)GetProcAddress(g_WtsApi32Module, "WT" \
+	WtsApi32_WtsApiFunctionTable.p##_name = (_type)GetProcAddress(g_WtsApi32Module, "WT" \
 	                                                                                  "S" #_name);
 
 static BOOL WtsApi32_InitializeWtsApi(void)
